### DISCUSSION TITLE
IRQ8 accuracy problem

### DISCUSSION THREAD

**Discussion by pif17** [dosemu2/dosemu2#2127] [Q&A]
Hello,

I would like to run a program compiled with DJGPP that uses IRQ8 at a frequency of 1000hz. 

I've done some tests but it's not really convincing in terms of repeatability, despite the use of a real-time kernel (PREMPT_RT patch).

Is there any way of achieving this with the Dosemu2 configuration?

Do I need to modify the source code to do this, and if so, where do I start?

I know that the program calls __dpmi_set_protected_mode_interrupt_vector(), how does Dosemu2 react to this? Is this the right place to start?

**Comment by stsp** [comment]
Well, when you say a "problem",
would you mind to describe the problem?

**Reply by pif17** [reply]
In this program, IRQ 8 is used to have a regular periodic interrupt every 1000µs.
With Dosemu2, the time measured between 2 executions varies from 85µs to 11259µs, with an average of 1039µs.
I'm looking for an average closer to 1000µs and a tighter amplitude. Let's say between 750µs and 1250µs.
I'm taking a closer look at how Dosemu2 handles IRQs, assuming that this repeatability problem lies here and not in KVM. I understand that DosEmu2 can delegate the interrupt table to KVM.

**Reply by stsp** [reply]
Yes, so far irq8 is guaranteed to
match the freq only on average.
I.e. if you measure for a few seconds,
you'll see the amount of IRQs is
overall correct, but on the short
runs frequency drifts a lot.

The reason is simple: IRQ8 was not
ported on a new timer stack properly,
but IRQ0 (partially) was.
This is a generic problem with dosemu2:
it still has dosemu1's components glued
to the new core "somehow", as I don't
have time for the proper porting work.

That said, you can try your luck with
IRQ0 instead. It still have some drifts,
but it should be much better, and it will
show you how IRQ8 can behave after
some more porting work.

**Reply by pif17** [reply]
OK,
I took a look to use IRQ0 instead of 8.
I still have missed interruption but the minimal time between 2 executions is closer that I expected.

Apart from your time, what else is needed to implement IRQ 0 and 8 as you wish?

**Reply by stsp** [reply]
> I still have missed interruption but the minimal time between 2 executions is closer that I expected.

Can you be more specific?
The measuring data may be valuable.

> Apart from your time, what else is needed to implement IRQ 0 and 8 as you wish?

I don't quite understand.
If you are satisfied about IRQ0, then
I guess you only need to ask about IRQ8?
Or IRQ0 also gave you unsatisfactory results?
Please be more specific, what do you want
to happen with IRQ0, and what do you want
to happen with IRQ8, separately and exactly?

**Reply by stsp** [reply]
I mean, even though the IRQ0 was
only partially ported, the work was
the bare minimum needed exactly
to get the frequency more or less
stable. So if you are unsatisfied even
with IRQ0's frequency, then its a bit
more than to just "implement it as I
wish". This will mean that more investigations
are needed first. Because I had no plans
working on IRQ0's frequency, I thought
its acceptable now.

**Reply by pif17** [reply]
What do I want with IRQ ?
I want to launch a periodic task 
with an accurate frequency of 1000hz

Why IRQ 0 or 8?
It was 2 options of the program that I 
use to get the searched accurate frequency 
of 1000hz. It's 2 ways to do the same thing.

> The measuring data may be valuable.

Here's a summary of the data I measure, on 
the same hardware, only the OS changes 
(Linux/PREEMPT_RT or FreeDOS) and the 
DPMI server (Dosemu2 default one on 
Linux or CWSDPMI on Freedos).
Measured are: time between 2 interrupt 
startups and interrupt execution time.
[DosEmu2TestIRQ.txt](https://github.com/dosemu2/dosemu2/files/13526780/DosEmu2TestIRQ.txt)
I focus on average l time between 2 startups that 
have to be closer to 1000µs and stadart deviation 
that have to be the smallest possible.
For the moment,  interrupt execution 
time doesn't realy matter.
To summarize:
- IRQ8 has too high a frequency (X2 factor with real hardware)
- IRQ0 and IRQ8 vary too much for me


>  if you are unsatisfied even
with IRQ0's frequency, then its a bit
more than to just "implement it as I
wish". This will mean that more investigations
are needed first. Because I had no plans
working on IRQ0's frequency, I thought
its acceptable now.

I have begun to investigate, I clearly understand 
that it's not simple and my need is surely too specific.
If I asked how you wished to 
implement this, it's because I have litle time
to spend on and perhaps give some help.
(I don't come to cry because my favorite game
doesn't work)

**Reply by stsp** [reply]
> IRQ8 has too high a frequency (X2 factor with real hardware)

That's likely a bug that needs to be fixed.
It would help if you open a ticket for
this, and attach a test-case there.

> If I asked how you wished to
> implement this, it's because I have litle time
> to spend on and perhaps give some help.

But IRQ0 is already implemented on top
of the new timer stack. It all starts in
lib/timer/evtimer_fd.c. It is supposed
to generate the call-backs with the needed
frequency, so you can start your measurements
from there.
Next interesting check-point is
misc/timers.c:timer_irq_ack().
It signals the delivery of an irq to
DOS, and also it manipulates the
counter pit[0].q_ticks. Which is an
amount of "queued" ticks (not yet
delivered to DOS). This var should
float around 0. If it gets increased,
then there are stalls somewhere.
